### PR TITLE
webpack5 support continued

### PR DESCRIPTION
**Pull request by aruniverse**
- Updated core-webpack-tools and backend-webpack-tools to support webpack 5.
	- now requires a peer dep of webpack 5 
	- plugins have been updated to support webpack 5
	- as of 3.0 we were missing support for keytar as an external so now electron apps should be able to webpack their backends
- See webpack 5 [changelog](https://github.com/webpack/changelog-v5/blob/master/README.md) and [migration guide](https://github.com/webpack/changelog-v5/blob/master/MIGRATION%20GUIDE.md)

Fixes https://github.com/iTwin/itwinjs-backlog/issues/234

### CODE REVIEW

File: docs/changehistory/NextVersion.md [hunk 1]
```diff
@@ -76,6 +79,10 @@ Reduced API surface of an `ElectronApp` class to only allow white-listed APIs fr
 
 The argument for [IModelHost.startup]($backend) has been changed from [IModelHostConfiguration]($backend) to the [IModelHostOptions]($backend) interface. This matches the approach on the frontend for [IModelApp.startup]($frontend) and makes it easier to supply startup options. `IModelHostConfiguration` implements `IModelHostOptions`, so existing code will continue to work without changes.
 
+## Tooling
+
+The `@itwin/core-webpack-tools` and `@itwin/backend-webpack-tools` packages have been updated to support [Webpack 5](https://webpack.js.org/) and now requires a peer dependency of _webpack@^5_. Please refer to their [changelog](https://github.com/webpack/changelog-v5/blob/master/README.md) and [migration guide](https://github.com/webpack/changelog-v5/blob/master/MIGRATION%20GUIDE.md) as you update. If you would like to continue using _webpack@^4_, please use versions ~3.2.0 of these packages.
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:84]: /s/requires/require

Comment (pmconne) [inline docs/changehistory/NextVersion.md:84]: > If you would like to continue using _webpack@^4_, please use versions ~3.2.0 of these packages

"These packages" meaning "iTwin.js packages"? Do we really want to encourage people not to update?

File: .vscode/settings.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
 {
   "editor.tabSize": 2,
   "editor.insertSpaces": true,
-  "editor.formatOnSave": true,
+  // "editor.formatOnSave": true,
```
Comment (pmconne) [inline .vscode/settings.json:5]: Were these edits to settings.json intentional?

File: docs/changehistory/NextVersion.md (hunk 1 above)
Comment (aruniverse) [reply docs/changehistory/NextVersion.md:84]:
```suggestion
The `@itwin/core-webpack-tools` and `@itwin/backend-webpack-tools` packages have been updated to support [Webpack 5](https://webpack.js.org/) and now require a peer dependency of _webpack@^5_. Please refer to their [changelog](https://github.com/webpack/changelog-v5/blob/master/README.md) and [migration guide](https://github.com/webpack/changelog-v5/blob/master/MIGRATION%20GUIDE.md) as you update. If you would like to continue using _webpack@^4_, please use versions ~3.2.0 of these packages.
```

File: .vscode/settings.json (hunk 2 above)
Comment (aruniverse) [reply .vscode/settings.json:5]: sorry will revert...

File: core/electron/src/frontend/ElectronApp.ts [hunk 3]
```diff
@@ -5,32 +5,40 @@
...
 /**
  * Frontend Ipc support for Electron apps.
  */
 class ElectronIpc implements IpcSocketFrontend {
-  private _api: ITwinElectronApi;
+  private _api: ITwinElectronApi | IpcRenderer;
   public addListener(channelName: string, listener: IpcListener) {
-    this._api.addListener(channelName, listener);
-    return () => this._api.removeListener(channelName, listener);
+    this._api?.addListener(channelName, listener);
```
Comment (pmconne) [inline core/electron/src/frontend/ElectronApp.ts:25]: I don't understand why the `?` was introduced - can `_api: ITwinElectronApi | IpcRenderer` ever be null or undefined?

File: docs/changehistory/NextVersion.md (hunk 1 above)
Comment (aruniverse) [reply docs/changehistory/NextVersion.md:84]: yes meaning the itwinjs pkgs, i will make that more explicit. 

Updating to webpack 5 is breaking change and imo these pkgs shouldve been broken out into their own version policy.
So it's not that we don't want users to update, it just might take them some effort and won't be as simple as just updating the pkg ver in the pkg.json. I could also remove this line altogether if you think thats best

Comment (pmconne) [reply docs/changehistory/NextVersion.md:84]: Removal seems preferable to me -- the change log is about what changed in 3.3, so it should be obvious to readers that if they don't like something that changed in 3.3, they can remain on 3.2.

File: core/electron/src/frontend/ElectronApp.ts (hunk 3 above)
Comment (aruniverse) [reply core/electron/src/frontend/ElectronApp.ts]: sorry this was left over from debugging, removed

File: docs/changehistory/NextVersion.md (hunk 1 above)
Comment (aruniverse) [reply docs/changehistory/NextVersion.md]: updated

Comment (aruniverse) [reply docs/changehistory/NextVersion.md]: updated

**Comment by pmconne** [comment]
display-test-app no longer displays annoying errors in electron window on startup (yay!) but still outputs warnings about webpack 5 and polyfills in terminal for stream and os. Should that be resolved?

**Comment by aruniverse** [comment]
> display-test-app no longer displays annoying errors in electron window on startup (yay!) but still outputs warnings about webpack 5 and polyfills in terminal for stream and os. Should that be resolved?

Those can be safely ignored. Will be fixed in 3.4 with changes from @swbsi new rpc pkgs.

**Review by pmconne — APPROVED** [review]

File: common/changes/@itwin/core-backend/webpack5-continued_2022-06-29-16-20.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/core-backend",
+      "comment": "Upgrade to Webpack 5",
```
Comment (calebmshafer) [inline common/changes/@itwin/core-backend/webpack5-continued_2022-06-29-16-20.json:5]: Can you drop the comment here?

File: core/electron/src/frontend/ElectronApp.ts [hunk 5]
```diff
@@ -5,15 +5,22 @@
...
+    itwinjs: ITwinElectronApi;
+  }
+}
 
 /**
  * Frontend Ipc support for Electron apps.
  */
 class ElectronIpc implements IpcSocketFrontend {
-  private _api: ITwinElectronApi;
+  private _api: ITwinElectronApi | IpcRenderer;
```
Comment (calebmshafer) [inline core/electron/src/frontend/ElectronApp.ts:23]: Why did you need to IpcRenderer here?

Comment (aruniverse) [reply core/electron/src/frontend/ElectronApp.ts:23]: thats the type when window.itwinjs doesn't exist,

```ts
this._api = window.itwinjs ?? require("electron").ipcRenderer;
```

Comment (calebmshafer) [reply core/electron/src/frontend/ElectronApp.ts:23]: ah okay, missed that before.

**Comment by aruniverse** [comment]
@pmconne tests on windows & linux are failing due to a circular reference in core-frontend/graphics caught in core-fullstack-tests

```
TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Batch'
    |     property 'perTargetData' -> object with constructor 'PerTargetData'
    --- property '_batch' closes the circle
    at JSON.stringify (<anonymous>)
    at Function.serialize (http://localhost:3010/src/runners/chrome/MochaSerializer.ts:65:17)
    at __puppeteer_evaluation_script__:11:120
    at Array.map (<anonymous>)
    at Runner.<anonymous> (__puppeteer_evaluation_script__:11:93)
    at emitTwo (http://localhost:3010/@/D:/vsts_a/6/s/common/temp/node_modules/.pnpm/mocha@10.0.0/node_modules/mocha/mocha.js:351:22)
    at Runner.emit (http://localhost:3010/@/D:/vsts_a/6/s/common/temp/node_modules/.pnpm/mocha@10.0.0/node_modules/mocha/mocha.js:425:9)
    at Runner.fail (http://localhost:3010/@/lib/runner.js:453:8)
    at http://localhost:3010/@/lib/runner.js:814:18
    at done (http://localhost:3010/@/lib/runnable.js:310:5)
```

See [Batch](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/render/webgl/Graphic.ts#L209) and [PerTargetData](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/render/webgl/Graphic.ts#L149). Mind taking a look?

I'm not yet entirely sure why this is only being caught now and causing issues only in the CI

**Comment by pmconne** [comment]
> @pmconne tests on windows & linux are failing due to a circular reference in core-frontend/graphics caught in core-fullstack-tests

They're not failing because of a circular reference. They're failing for some other reason, and mocha is attempting to report the error but tripping over its feet by trying to JSON.stringify an object containing circular references. You'll have to break in debugger to see the actual test failure.

**Comment by aruniverse** [comment]
> They're failing for some other reason

Looks like it was an issue with the webpack config not merging configs together, tests are now passing.

This pr should be g2g, just need review/approval from @calebmshafer or @wgoehrig

File: docs/changehistory/NextVersion.md [hunk 6]
```diff
@@ -8,17 +8,23 @@ Table of contents:
 - [Display system](#display-system)
   - [Dynamic schedule scripts](#dynamic-schedule-scripts)
   - [Hiliting models and subcategories](#hiliting-models-and-subcategories)
+- [Frontend category APIs](#frontend-category-apis)
```
Comment (calebmshafer) [inline docs/changehistory/NextVersion.md:11]: is this supposed to be indented this way?

File: docs/changehistory/NextVersion.md [hunk 7]
```diff
@@ -8,17 +8,23 @@ Table of contents:
...
-  - [Tool Settings title](tool-settings-title)
+  - [Tool Settings title](#tool-settings-title)
 - [ElectronApp changes](#electronapp-changes)
 - [IModelHostOptions](#imodelhostoptions)
 - [Progress API for downloading changesets](#progress-api-for-downloading-changesets)
-- [Deprecations](#deprecations)
+- [RPC over IPC](#rpc-over-ipc)
 - [Presentation](#presentation)
+  - [Relationship properties](#relationship-properties)
+- [Tooling](#tooling)
```
Comment (calebmshafer) [inline docs/changehistory/NextVersion.md:23]: Let's call this out as webpack 5 specifically in the TOC, that way it doesn't get lost when people review it.

File: docs/changehistory/NextVersion.md (hunk 6 above)
Comment (aruniverse) [reply docs/changehistory/NextVersion.md:11]: i believe so, this was autocorrected by my md vscode extensions.
this is the parent header for the ones below

see, https://github.com/iTwin/itwinjs-core/blob/webpack5-continued/docs/changehistory/NextVersion.md#frontend-category-apis and look at the display rendered blob 
image

File: docs/changehistory/NextVersion.md (hunk 7 above)
Comment (aruniverse) [reply docs/changehistory/NextVersion.md:23]:
```suggestion
- [Webpack 5](#webpack-5)
```

File: docs/changehistory/NextVersion.md (hunk 6 above)
Comment (pmconne) [reply docs/changehistory/NextVersion.md:11]: No, it's out of order. I'll fix.

Comment (pmconne) [reply docs/changehistory/NextVersion.md]: Fixed.

**Review by calebmshafer — APPROVED** [review]
LGTM. I did not closely review the webpack plugin changes that closely, everything else looks good. Hopefully @wgoehrig will have a chance to take a look.

What is our timeframe for update react-scripts to 5? We'll want to update our changelog once that happens so consumers know to use it.

**Comment by aruniverse** [comment]
> What is our timeframe for update react-scripts to 5? 

We first need to publish the 3.3 pkgs. I guess technically could pull in a dev-release into cra but not a huge fan of that. Can update changelogs in 3.3.1 with a mention of it. Do we need to explitly call out our cra5 fork here in the core pkgs/changelogs?

**Comment by calebmshafer** [comment]
> > What is our timeframe for update react-scripts to 5?
> 
> We first need to publish the 3.3 pkgs. I guess technically could pull in a dev-release into cra but not a huge fan of that. Can update changelogs in 3.3.1 with a mention of it. Do we need to explitly call out our cra5 fork here in the core pkgs/changelogs?

@aruniverse what about pulling in the RC we are hopefully going to have later today and release a new version of `@bentley/react-scripts`?

**Review by wgoehrig — APPROVED** [review]

File: tools/webpack-core/src/utils/IModelJsOptionsDefaulter.ts [hunk 8]
```diff
@@ -2,53 +2,15 @@
...
-          test: /\.js$/,
-          loader: path.join(__dirname, "../loaders/strip-assert-loader.js"),
-          enforce: "pre",
-        });
-
-      return [
-        ...rulesToAdd, ...(value || []),
-      ];
-    });
+  public override process(_options: WebpackOptionsNormalized) {
```
Comment (wgoehrig) [inline tools/webpack-core/src/utils/IModelJsOptionsDefaulter.ts:13]: Let's just delete this whole file (can be in a future PR)

**Comment by aruniverse** [comment]
Force merged as all requirements were satisfied except the cla which im pretty sure ive signed at some point :)

image
